File: 38.外观数列.ts
/*
 * @lc app=leetcode.cn id=38 lang=typescript
 *
 * [38] 外观数列
 */

// @lc code=start
function countAndSay(n: number): string {
    let ret = "1";
    for (let i = 2; i <= n; ++i) {
        let [p, start, tmp]: [number, number, string[]] = [0, 0, []];
        while (p < ret.length) {
            for (;p<ret.length && ret[p]===ret[start];++p) {}
            tmp.push(''+(p-start)+ret[start]);
            start = p;
        }
        ret = tmp.join("");
    }
    return ret;
};
// @lc code=end

File: 38.外观数列.rs
/*
 * @lc app=leetcode.cn id=38 lang=rust
 *
 * [38] 外观数列
 */

// @lc code=start
impl Solution {
    pub fn count_and_say(n: i32) -> String {
        let mut ret = vec!['1'];
        for i in 2..=n {
            let (mut p, mut start) = (0, 0);
            let mut tmp = Vec::new();
            while p < ret.len() {
                while p < ret.len() && ret[p] == ret[start] {
                    p += 1;
                }
                tmp.extend((p - start).to_string().chars().collect::<Vec<char>>());
                tmp.push(ret[start]);
                start = p;
            }
            ret = tmp;
        }
        ret.iter().collect()
    }
}
// @lc code=end
